crates/rrg/src/action/get_filesystem_timeline_tsk.rs

```diff
@@ -0,0 +1,582 @@
+pub fn get_mount(mounts: &[Mount], path: &Path) -> std::io::Result<Mount> {
+    // Note: std::path::absolute does not follow symlinks like std::fs::canonicalize does.
+    // It also returns C:\-style paths on windows rather than \\?\C:\-style paths.
+    let path: PathBuf = std::path::absolute(path)?;
+    mounts
+        .iter()
+        .cloned()
+        .filter_map(|mut m| {
+            std::path::absolute(&m.path).ok().map(move |abs_path| {
+                m.path = abs_path;
+                m
+            })
+        })
+        .filter(|m| path.starts_with(&m.path))
+        // Filter out dummy and remote filesystems whose names may start with /.
+        .max_by_key(|m| m.path.as_os_str().len())
+        .ok_or_else(|| {
+            std::io::Error::new(
+                std::io::ErrorKind::NotFound,
+                format!("Failed to locate mount for {path:?}"),
+            )
+        })
+}
+
+pub fn get_raw_device(mounts: &[Mount], path: &Path) -> std::io::Result<RawDeviceMount> {
+    let mount = get_mount(mounts, path)?;
+    #[cfg(not(target_os = "windows"))]
+    {
+        // Comprehensive list of raw filesystem types supported by RRG.
+        const SUPPORTED_FS_TYPES: &[&str] = &["ext2", "ext3", "ext4", "vfat", "ntfs", "fuseblk"];
+        if !SUPPORTED_FS_TYPES.contains(&mount.fs_type.as_ref()) {
+            return Err(std::io::Error::other(format!(
+                "Unsupported filesystem type: {}",
+                mount.fs_type
+            )));
+        }
+        Ok(RawDeviceMount {
+            image_path: mount.name.into(),
+            mountpoint: mount.path.to_path_buf(),
+        })
+    }
+    #[cfg(target_os = "windows")]
+    {
+        Ok(RawDeviceMount {
+            // The "name" of the mount is the volume GUID path: \\?\Volume{...}\
+            // Opening \\?\Volume{...}\ opens the directory.
+            // Opening \\?\Volume{...} opens the raw bytes of the volume.
+            image_path: mount.name.trim_end_matches('\\').into(),
+            mountpoint: mount.path.to_path_buf(),
+        })
+    }
+}
+
+/// Given an absolute path and the raw device that contains it, computes the path
+/// embedded within the raw device to give TSK.
+///
+/// This path must start with `/`, and this function normalizes all path
+/// separators in the embedded path to use `/` for consistency.
+fn get_embedded_path(mount: &RawDeviceMount, absolute_path: &Path) -> PathBuf {
+    absolute_path
+        .strip_prefix(&mount.mountpoint)
+        .expect("mountpoint not prefix of root")
+        .iter()
+        .flat_map(|component| [OsStr::new("/"), component])
+        .collect::<OsString>()
+        .into()
+}
+
+/// Handles requests for the timeline action.
+pub fn handle<S>(session: &mut S, args: Args) -> crate::session::Result<()>
+where
+    S: crate::session::Session,
+{
+    use sha2::Digest as _;
+
+    let raw_fs: PathBuf;
+    let embedded_path;
+    if !args.raw_fs.as_os_str().is_empty() {
```

Comment:
It would make `args.raw_fs` optional, do the empty-to-optional conversion at the `from_proto` level and then use pattern matching here.